app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php

```diff
@@ -1793,4 +2054,398 @@ protected function _getDeliveryConfirmationLevel($countyDest = null)
+                        "Address" => [
+                            "AddressLine" => ["{$params['49_residential']}"],
+                            "StateProvinceCode" => "{$params['destRegionCode']}",
+                            "PostalCode" => "{$params['19_destPostal']}",
+                            "CountryCode" => "{$params['22_destCountry']}",
+                            "ResidentialAddressIndicator" => "{$residentialAddressIndicator}"
+                        ]
+                    ],
+                    "ShipFrom" => [
+                        "Address" => [
+                            "AddressLine" => [],
+                            "StateProvinceCode" => "{$params['origRegionCode']}",
+                            "PostalCode" => "{$params['15_origPostal']}",
+                            "CountryCode" => "{$params['14_origCountry']}"
+                        ]
+                    ],
+                ]
+            ]
+        ];
+
+        if ($this->getConfigFlag('negotiated_active')) {
+            $rateParams['RateRequest']['Shipment']['ShipmentRatingOptions']['TPFCNegotiatedRatesIndicator'] = "Y";
+            $rateParams['RateRequest']['Shipment']['ShipmentRatingOptions']['NegotiatedRatesIndicator'] = "Y";
+        }
+        if ($this->getConfigFlag('include_taxes')) {
+            $rateParams['RateRequest']['Shipment']['TaxInformationIndicator'] = "Y";
+        }
+
+        if ($serviceCode !== null) {
+            $rateParams['RateRequest']['Shipment']['Service']['code'] = $serviceCode;
+            $rateParams['RateRequest']['Shipment']['Service']['Description'] = $serviceDescription;
+        }
+
+        foreach ($rowRequest->getPackages() as $package) {
+            $rateParams['RateRequest']['Shipment']['Package'][] = [
+                "PackagingType" => [
+                    "Code" => "{$params['48_container']}",
+                    "Description" => "Packaging"
+                ],
+                "Dimensions" => [
+                    "UnitOfMeasurement" => [
+                        "Code" => "IN",
+                        "Description" => "Inches"
+                    ],
+                    "Length" => "5",
+                    "Width" => "5",
+                    "Height" => "5"
+                ],
+                "PackageWeight" => [
+                    "UnitOfMeasurement" => [
+                        "Code" => "{$rowRequest->getUnitMeasure()}"
+                    ],
+                    "Weight" => "{$this->_getCorrectWeight($package['weight'])}"
+                ]
+            ];
+        }
+
+        $ratePayload = json_encode($rateParams, JSON_PRETTY_PRINT);
+        /** Rest API Payload */
+        $version = "v1";
+        $requestOption = $params['10_action'];
+        $headers = [
+            "Authorization: Bearer {$accessToken}",
+            "Content-Type: application/json"
+        ];
+        $debugData = [
+            'request' => $ratePayload
+        ];
+
+        try {
```

Comment:
This is also something I'm trying to wrap my head around. Why are the `curl_*()` functions wrapped in a try-catch? These functions won't *ever* throw an Exception, AFAIK.

There are two cases where it might throw an `Error`, but it should never throw an `Exception`. If you pass an invalid url to `curl_init()`, then it *might* return `false` (though I've never seen it do that, even if you pass in total garbage); if you *then* pass that `false` into `curl_setopt` or `curl_exec`, then you might see a thrown `TypeError`. Or, if you try to call any of the `curl_*` functions when the php-curl extension isn't loaded, you'll also get a `FatalError` thrown.

But neither of those are subclasses of `Exception`, so the try-catches really don't make any sense to me. Am I missing something?

For error handling the cURL functions, we should be checking `curl_errno()` and `curl_error()`, or at least checking if `$responseData === false`.

Reply:
I'll remove the try/catch but it was already used in the XML implementation of the module